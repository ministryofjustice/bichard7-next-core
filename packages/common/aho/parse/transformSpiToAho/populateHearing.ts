import type { Hearing } from "../../../types/AnnotatedHearingOutcome"
import type { ResultedCaseMessageParsedXml, SpiDefendant } from "../../../types/SpiResult"

import getOrganisationUnit from "../../getOrganisationUnit"
import removeSeconds from "./removeSeconds"

type DefendantDetailsData = {
  name: string
  presentAtHearing: string
}

const getDefendantDetails = (spiDefendant: SpiDefendant): DefendantDetailsData => {
  if (spiDefendant.CourtIndividualDefendant) {
    const {
      CourtIndividualDefendant: {
        PersonDefendant: {
          BasePersonDetails: {
            PersonName: { PersonFamilyName: familyName, PersonGivenName1: givenName }
          }
        },
        PresentAtHearing: presentAtHearing
      }
    } = spiDefendant

    const name = [givenName, familyName]
      .filter((namePart) => namePart)
      .join(" ")
      .trim()
    return { name, presentAtHearing }
  } else if (spiDefendant.CourtCorporateDefendant) {
    const {
      CourtCorporateDefendant: {
        OrganisationName: { OrganisationName: spiOrganisationName },
        PresentAtHearing: presentAtHearing
      }
    } = spiDefendant
    const name = spiOrganisationName.trim()
    return { name, presentAtHearing }
  }

  throw new Error("Defendant details contained neither CourtIndividualDefendant or CourtCorporateDefendant")
}

const populateHearing = (messageId: string, courtResult: ResultedCaseMessageParsedXml): Hearing => {
  const {
    Session: {
      Case: { Defendant: spiDefendant },
      CourtHearing: {
        Hearing: {
          CourtHearingLocation: spiCourtHearingLocation,
          DateOfHearing: spiDateOfHearing,
          TimeOfHearing: spiTimeOfHearing
        },
        PSAcode: spiPsaCode
      }
    }
  } = courtResult

  const { name, presentAtHearing } = getDefendantDetails(spiDefendant)

  const hearingOutcomeHearing: Hearing = {
    CourtHearingLocation: getOrganisationUnit(spiCourtHearingLocation),
    CourtHouseCode: Number(spiPsaCode),
    DateOfHearing: new Date(spiDateOfHearing),
    DefendantPresentAtHearing: presentAtHearing,
    HearingDocumentationLanguage: "D",
    HearingLanguage: "D",
    SourceReference: {
      DocumentName: `SPI ${name}`,
      DocumentType: "SPI Case Result",
      UniqueID: messageId
    },
    TimeOfHearing: removeSeconds(spiTimeOfHearing)
  }

  return hearingOutcomeHearing
}

export default populateHearing
