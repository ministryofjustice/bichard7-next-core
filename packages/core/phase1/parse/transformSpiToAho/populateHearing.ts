import type { Hearing } from "../../../types/AnnotatedHearingOutcome"
import getOrganisationUnit from "../../lib/organisationUnit/getOrganisationUnit"
import type { ResultedCaseMessageParsedXml, SpiDefendant } from "../../types/SpiResult"
import removeSeconds from "./removeSeconds"

type DefendantDetailsData = {
  presentAtHearing: string
  name: string
}

const getDefendantDetails = (spiDefendant: SpiDefendant): DefendantDetailsData => {
  if (spiDefendant.CourtIndividualDefendant) {
    const {
      CourtIndividualDefendant: {
        PresentAtHearing: presentAtHearing,
        PersonDefendant: {
          BasePersonDetails: {
            PersonName: { PersonGivenName1: givenName, PersonFamilyName: familyName }
          }
        }
      }
    } = spiDefendant

    const name = [givenName, familyName]
      .filter((namePart) => namePart)
      .join(" ")
      .trim()
    return { presentAtHearing, name }
  } else if (spiDefendant.CourtCorporateDefendant) {
    const {
      CourtCorporateDefendant: {
        OrganisationName: { OrganisationName: spiOrganisationName },
        PresentAtHearing: presentAtHearing
      }
    } = spiDefendant
    const name = spiOrganisationName.trim()
    return { presentAtHearing, name }
  }
  throw new Error("Defendant details contained neither CourtIndividualDefendant or CourtCorporateDefendant")
}

const populateHearing = (messageId: string, courtResult: ResultedCaseMessageParsedXml): Hearing => {
  const {
    Session: {
      CourtHearing: {
        Hearing: {
          CourtHearingLocation: spiCourtHearingLocation,
          DateOfHearing: spiDateOfHearing,
          TimeOfHearing: spiTimeOfHearing
        },
        PSAcode: spiPsaCode
      },
      Case: { Defendant: spiDefendant }
    }
  } = courtResult

  const { presentAtHearing, name } = getDefendantDetails(spiDefendant)

  const hearingOutcomeHearing: Hearing = {
    CourtHearingLocation: getOrganisationUnit(spiCourtHearingLocation),
    DateOfHearing: new Date(spiDateOfHearing),
    TimeOfHearing: removeSeconds(spiTimeOfHearing),
    HearingLanguage: "D",
    HearingDocumentationLanguage: "D",
    CourtHouseCode: Number(spiPsaCode),
    SourceReference: {
      DocumentName: `SPI ${name}`,
      DocumentType: "SPI Case Result",
      UniqueID: messageId
    },
    DefendantPresentAtHearing: presentAtHearing
  }

  return hearingOutcomeHearing
}

export default populateHearing
