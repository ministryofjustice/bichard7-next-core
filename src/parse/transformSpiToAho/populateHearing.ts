import getOrganisationUnit from "../../lib/organisationUnit/getOrganisationUnit"
import type { Hearing } from "../../types/AnnotatedHearingOutcome"
import type { ResultedCaseMessageParsedXml } from "../../types/SpiResult"
import removeSeconds from "./removeSeconds"

export default (messageId: string, courtResult: ResultedCaseMessageParsedXml): Hearing => {
  const {
    Session: {
      CourtHearing: {
        Hearing: {
          CourtHearingLocation: spiCourtHearingLocation,
          DateOfHearing: spiDateOfHearing,
          TimeOfHearing: spiTimeOfHearing
        },
        PSAcode: spiPsaCode
      },
      Case: { Defendant: spiDefendant }
    }
  } = courtResult

  const hearingOutcomeHearing = {} as Hearing

  hearingOutcomeHearing.CourtHearingLocation = getOrganisationUnit(spiCourtHearingLocation)

  hearingOutcomeHearing.DateOfHearing = new Date(spiDateOfHearing)
  hearingOutcomeHearing.TimeOfHearing = removeSeconds(spiTimeOfHearing)
  hearingOutcomeHearing.HearingLanguage = "D"
  hearingOutcomeHearing.HearingDocumentationLanguage = "D"

  let name = ""
  if (spiDefendant.CourtIndividualDefendant) {
    const {
      CourtIndividualDefendant: {
        PresentAtHearing: spiPresentAtHearing,
        PersonDefendant: {
          BasePersonDetails: {
            PersonName: { PersonGivenName1: givenName, PersonFamilyName: familyName }
          }
        }
      }
    } = spiDefendant

    hearingOutcomeHearing.DefendantPresentAtHearing = spiPresentAtHearing
    name = [givenName, familyName]
      .filter((namePart) => namePart)
      .join(" ")
      .trim()
  } else if (spiDefendant.CourtCorporateDefendant) {
    const {
      CourtCorporateDefendant: {
        OrganisationName: { OrganisationName: spiOrganisationName },
        PresentAtHearing: spiPresentAtHearing
      }
    } = spiDefendant
    name = spiOrganisationName
    hearingOutcomeHearing.DefendantPresentAtHearing = spiPresentAtHearing
  }
  hearingOutcomeHearing.SourceReference = {
    DocumentName: `SPI ${name}`,
    DocumentType: "SPI Case Result",
    UniqueID: messageId
  }

  hearingOutcomeHearing.CourtHouseCode = Number(spiPsaCode)

  return hearingOutcomeHearing
}
