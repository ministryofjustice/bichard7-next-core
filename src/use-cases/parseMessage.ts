import { XMLParser } from "fast-xml-parser"
import type { ResultedCaseMessageParsedXml } from "src/types/IncomingMessage"
import { incomingMessageParsedXmlSchema } from "src/types/IncomingMessage"

export default (message: string): ResultedCaseMessageParsedXml => {
  const options = {
    ignoreAttributes: false,
    removeNSPrefix: true
  }

  const parser = new XMLParser(options)
  const rawParsedObj = parser.parse(message) as unknown
  const parsedObj = incomingMessageParsedXmlSchema.parse(rawParsedObj)
  const rcm = parsedObj.DeliverRequest.Message.ResultedCaseMessage
  // if (!rcm.Session.Case.Defendant.Offence) {
  //   rcm.Session.Case.Defendant.Offence = []
  // }
  // if (!Array.isArray(rcm.Session.Case.Defendant.Offence)) {
  //   rcm.Session.Case.Defendant.Offence = [rcm.Session.Case.Defendant.Offence]
  // }
  // rcm.Session.Case.Defendant.Offence.forEach((offence) => {
  //   if (!offence.Result) {
  //     offence.Result = []
  //   }
  //   if (!Array.isArray(offence.Result)) {
  //     offence.Result = [offence.Result]
  //   }
  // })
  return rcm
}
