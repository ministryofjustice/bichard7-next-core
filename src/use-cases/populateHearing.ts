import type { Hearing } from "src/types/AnnotatedHearingOutcome"
import type { ResultedCaseMessageParsedXml } from "src/types/IncomingMessage"
import removeSeconds from "src/utils/removeSeconds"
import getOrganisationUnit from "./getOrganisationUnit"

export default (messageId: string, courtResult: ResultedCaseMessageParsedXml): Hearing => {
  const {
    Session: {
      CourtHearing: {
        Hearing: {
          CourtHearingLocation: spiCourtHearingLocation,
          DateOfHearing: spiDateOfHearing,
          TimeOfHearing: spiTimeOfHearing
        },
        PSAcode: spiPsaCode
      },
      Case: { Defendant: spiDefendant }
    }
  } = courtResult

  const hearingOutcomeHearing = {} as Hearing

  hearingOutcomeHearing.CourtHearingLocation = getOrganisationUnit(spiCourtHearingLocation)

  hearingOutcomeHearing.DateOfHearing = spiDateOfHearing
  hearingOutcomeHearing.TimeOfHearing = removeSeconds(spiTimeOfHearing)
  hearingOutcomeHearing.HearingLanguage = "D"
  hearingOutcomeHearing.HearingDocumentationLanguage = "D"

  let name = ""
  if (spiDefendant.CourtIndividualDefendant) {
    const {
      CourtIndividualDefendant: {
        PresentAtHearing: spiPresentAtHearing,
        PersonDefendant: {
          BasePersonDetails: {
            PersonName: { PersonGivenName1: givenName, PersonFamilyName: familyName }
          }
        }
      }
    } = spiDefendant

    hearingOutcomeHearing.DefendantPresentAtHearing = spiPresentAtHearing
    name = [givenName, familyName].filter((namePart) => namePart).join(" ")
  } else if (spiDefendant.CourtCorporateDefendant) {
    name = spiDefendant.CourtCorporateDefendant.OrganisationName.OrganisationName
  }
  hearingOutcomeHearing.SourceReference = {
    DocumentName: `SPI ${name}`,
    DocumentType: "SPI Case Result",
    UniqueID: messageId
  }

  hearingOutcomeHearing.CourtHouseCode = spiPsaCode

  return hearingOutcomeHearing
}
