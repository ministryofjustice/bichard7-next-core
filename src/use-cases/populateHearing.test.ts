import { readFileSync } from "fs"
import parseMessage from "./parseMessage"
import populateHearing from "./populateHearing"

const message = readFileSync("test-data/input-message-001.xml", "utf-8")
const courtResult = parseMessage(message)

describe("populateHearing", () => {
  it("should transform SPI Hearing to Hearing Outcome Hearing", () => {
    const messageId = "DUMMY_MESSAGE_ID"
    const hearingOutcomeCase = populateHearing(messageId, courtResult)

    expect(hearingOutcomeCase).toBeDefined()
    expect(hearingOutcomeCase).toMatchSnapshot()
  })
})
