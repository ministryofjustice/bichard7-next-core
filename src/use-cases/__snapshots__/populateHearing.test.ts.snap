// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`populateHearing should transform SPI Hearing to Hearing Outcome Hearing 1`] = `
Object {
  "CourtHearingLocation": Object {
    "BottomLevelCode": "01",
    "OrganisationUnitCode": Object {
      "attributes": undefined,
      "value": "B01EF01",
    },
    "SecondLevelCode": "01",
    "ThirdLevelCode": "EF",
    "TopLevelCode": "B",
  },
  "CourtHouseCode": Object {
    "attributes": undefined,
    "value": 2576,
  },
  "DateOfHearing": Object {
    "attributes": undefined,
    "value": "2011-09-26",
  },
  "DefendantPresentAtHearing": Object {
    "attributes": undefined,
    "value": "A",
  },
  "HearingDocumentationLanguage": Object {
    "attributes": undefined,
    "value": "D",
  },
  "HearingLanguage": Object {
    "attributes": undefined,
    "value": "D",
  },
  "SourceReference": Object {
    "DocumentName": Object {
      "attributes": undefined,
      "value": "SPI TRPRFOUR SEXOFFENCE",
    },
    "DocumentType": Object {
      "attributes": undefined,
      "value": "SPI Case Result",
    },
    "UniqueID": Object {
      "attributes": undefined,
      "value": "DUMMY_MESSAGE_ID",
    },
  },
  "TimeOfHearing": Object {
    "attributes": undefined,
    "value": "10:00",
  },
}
`;
